File: common/src/main/java/me/pandamods/pandalib/api/config/holders/ConfigHolder.java
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package me.pandamods.pandalib.api.config.holders;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import dev.architectury.platform.Platform;
import me.pandamods.pandalib.PandaLib;
import me.pandamods.pandalib.api.config.Config;
import me.pandamods.pandalib.api.config.ConfigData;
import me.pandamods.pandalib.utils.ClassUtils;
import net.minecraft.resources.ResourceLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;

public class ConfigHolder<T extends ConfigData> {
	public final Logger logger;
	private final Gson gson;

	private final Class<T> configClass;
	private final Config definition;
	private final ResourceLocation resourceLocation;
	private final boolean synchronize;
	private T config;

	public ConfigHolder(Class<T> configClass, Config config) {
		this.configClass = configClass;
		this.definition = config;
		this.logger = LoggerFactory.getLogger(config.modId() + " | Config");
		this.gson = getNewDefault().buildGson(new GsonBuilder()).setPrettyPrinting().create();

		this.resourceLocation = ResourceLocation.fromNamespaceAndPath(config.modId(), config.name());
		this.synchronize = config.synchronize();

		if (this.load()) {
			save();
		}
	}

	public Gson getGson() {
		return gson;
	}

	public Class<T> getConfigClass() {
		return configClass;
	}

	public Config getDefinition() {
		return definition;
	}

	public boolean shouldSynchronize() {
		return synchronize;
	}

	public Path getConfigPath() {
		Path path = Platform.getConfigFolder();
		if (!definition.parentDirectory().isBlank()) path = path.resolve(definition.parentDirectory());
		return path.resolve(definition.name() + ".json");
	}

	public void save() {
		JsonObject jsonObject = this.getGson().toJsonTree(this.config).getAsJsonObject();
		this.config.onSave(this, jsonObject);
		Path configPath = getConfigPath();
		try {
			Files.createDirectories(configPath.getParent());
			BufferedWriter writer = Files.newBufferedWriter(configPath);
			this.getGson().toJson(jsonObject, writer);
			writer.close();
			this.logger.info("Successfully saved config '{}'", definition.name());
		} catch (IOException e) {
			this.logger.info("Failed to save config '{}'", definition.name());
			throw new RuntimeException(e);
		}
	}

	public boolean load() {
		Path configPath = getConfigPath();
		if (Files.exists(configPath)) {
			try (BufferedReader reader = Files.newBufferedReader(configPath)) {
				JsonObject jsonObject = this.getGson().fromJson(reader, JsonObject.class);
				this.config = this.getGson().fromJson(jsonObject, configClass);
				this.config.onLoad(this, jsonObject);
			} catch (IOException e) {
				this.logger.error("Failed to load config '{}', using default", definition.name(), e);
				resetToDefault();
				return false;
			}
		} else {
			resetToDefault();
			save();
		}
		this.logger.info("Successfully loaded config '{}'", definition.name());
		return true;
	}

	public void resetToDefault() {
		this.config = getNewDefault();
	}

	/**
	 * @return Newly created class
	 */
	public T getNewDefault() {
		return ClassUtils.constructUnsafely(configClass);
	}

	public ResourceLocation resourceLocation() {
		return resourceLocation;
	}

	public String getLangName() {
		return String.format("config.%s.%s", resourceLocation.getNamespace(), resourceLocation.getPath());
	}

	public String modID() {
		return getDefinition().modId();
	}

	/**
	 * @return Local config settings
	 */
	public T get() {
		return config;
	}
}

File: common/src/main/kotlin/me/pandamods/pandalib/config/holders/ConfigHolder.kt
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package me.pandamods.pandalib.config.holders

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import dev.architectury.platform.Platform
import me.pandamods.pandalib.config.Config
import me.pandamods.pandalib.config.ConfigData
import me.pandamods.pandalib.platform.Services
import me.pandamods.pandalib.utils.ClassUtils
import net.minecraft.resources.ResourceLocation
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path

open class ConfigHolder<T : ConfigData>(val configClass: Class<T>, val definition: Config) {
	val logger: Logger = LoggerFactory.getLogger(definition.modId + " | Config")
	val gson: Gson = newDefault.buildGson(GsonBuilder()).setPrettyPrinting().create()

	private val resourceLocation: ResourceLocation = ResourceLocation.fromNamespaceAndPath(definition.modId, definition.name)
	private val synchronize: Boolean = definition.synchronize
	private lateinit var config: T

	init {
		if (load()) {
			save()
		}
	}

	fun shouldSynchronize(): Boolean {
		return synchronize
	}

	val configPath: Path = Services.GAME.configDir.let {
		if (definition.directory.isBlank()) it else it.resolve(definition.directory)
	}.resolve(definition.name + ".json")

	open fun save() {
		val jsonObject = gson.toJsonTree(config).getAsJsonObject()
		this.config.onSave(this, jsonObject)
		
		val configPath = this.configPath
		try {
			Files.createDirectories(configPath.parent)
			val writer = Files.newBufferedWriter(configPath)
			this.gson.toJson(jsonObject, writer)
			writer.close()
			this.logger.info("Successfully saved config '{}'", definition.name)
		} catch (e: IOException) {
			this.logger.info("Failed to save config '{}'", definition.name)
			throw RuntimeException(e)
		}
	}

	open fun load(): Boolean {
		val configPath = this.configPath
		if (Files.exists(configPath)) {
			try {
				Files.newBufferedReader(configPath).use { reader ->
					val jsonObject = this.gson.fromJson<JsonObject>(reader, JsonObject::class.java)
					this.config = this.gson.fromJson(jsonObject, configClass)
					this.config.onLoad(this, jsonObject)
				}
			} catch (e: IOException) {
				this.logger.error("Failed to load config '{}', using default", definition.name, e)
				resetToDefault()
				return false
			}
		} else {
			resetToDefault()
			save()
		}
		this.logger.info("Successfully loaded config '{}'", definition.name)
		return true
	}

	fun resetToDefault() {
		this.config = this.newDefault
	}

	val newDefault: T
		/**
		 * @return Newly created class
		 */
		get() = ClassUtils.constructUnsafely(configClass)

	fun resourceLocation(): ResourceLocation {
		return resourceLocation
	}

	val langName: String
		get() = String.format("config.%s.%s", resourceLocation.namespace, resourceLocation.path)

	fun modID(): String {
		return this.definition.modId
	}

	/**
	 * @return Local config settings
	 */
	open fun get(): T? {
		return config
	}
}
